javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -561,20 +566,22 @@ private Vector(int indexShift, HashArrayMappedTrie<Integer, T> trie) {
 
     @Override
     public Vector<T> append(T element) {
-        if (isEmpty()) {
-            return of(element);
-        } else {
-            return new Vector<>(indexShift, trie.put(length() + indexShift, element));
-        }
+        return wrap(trie.appendLeaf(element));
     }
 
     @Override
     public Vector<T> appendAll(Iterable<? extends T> elements) {
-        HashArrayMappedTrie<Integer, T> result = trie;
-        for (T element : elements) {
-            result = result.put(result.size() + indexShift, element);
+        Objects.requireNonNull(elements, "elements is null");
+        if (isEmpty()) {
+            return ofAll(elements);
+        } else {
+            Vector<T> result = this;
+            for (T element : elements) {
+                result = result.append(element);
+            }
+            return result.isEmpty() ? empty()
+                                    : result;
         }
-        return new Vector<>(indexShift, result);
```

Comment:
Great! Just to get sure - `history.forEach(...)` also checks that previous versions remain unmodified? I.e. `t._1` is independent of `t._2`